parser-typechecker/src/Unison/Util/Text.hs

```diff
@@ -0,0 +1,133 @@
+{-# Language BangPatterns #-}
+{-# Language GeneralizedNewtypeDeriving #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Unison.Util.Text where
+
+import Data.String (IsString(..))
+import Data.Foldable (toList)
+import Data.List (unfoldr)
+import Prelude hiding (take,drop,replicate)
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Unison.Util.Bytes as B
+import qualified Unison.Util.Rope as R
+
+-- Text type represented as a `Rope` of chunks
+newtype Text = Text (R.Rope Chunk) deriving (Eq,Ord,Semigroup,Monoid)
+
+data Chunk = Chunk {-# unpack #-} !Int {-# unpack #-} !T.Text
+
+empty :: Text
+empty = Text mempty
+
+one, singleton :: Char -> Text
+one ch = Text (R.one (chunk (T.singleton ch)))  
+singleton = one
+
+threshold :: Int
+threshold = 512
+
+replicate :: Int -> Text -> Text
+replicate n t | size t * n < threshold = Text (R.one (chunk (T.replicate n (toText t)))) 
+replicate 0 _ = mempty 
+replicate 1 t = t
+replicate n t = 
+  replicate (n `div` 2) t <> replicate (n - (n `div` 2)) t
+
+chunkToText :: Chunk -> T.Text
+chunkToText (Chunk _ t) = t
+
+chunk :: T.Text -> Chunk
+chunk t = Chunk (T.length t) t 
+
+take :: Int -> Text -> Text
+take n (Text t) = Text (R.take n t)
+
+drop :: Int -> Text -> Text
+drop n (Text t) = Text (R.drop n t)
+
+uncons :: Text -> Maybe (Char, Text)
+uncons t | size t == 0 = Nothing
+uncons t = (,drop 1 t) <$> at 0 t
+
+unsnoc :: Text -> Maybe (Text, Char)
+unsnoc t | size t == 0 = Nothing
+unsnoc t = (take (size t - 1) t,) <$> at (size t - 1) t
+
+at :: Int -> Text -> Maybe Char
+at n (Text t) = R.index n t
+
+size :: Text -> Int
+size (Text t) = R.size t
+
+reverse :: Text -> Text
+reverse (Text t) = Text (R.reverse t)
+
+fromUtf8 :: B.Bytes -> Either String Text
+fromUtf8 bs = 
+  case T.decodeUtf8' (B.toByteString bs) of
+    Right t -> Right (fromText t)
+    Left e -> Left (show e)
+
+toUtf8 :: Text -> B.Bytes
+toUtf8 (Text t) = B.Bytes (R.map (B.chunkFromByteString . T.encodeUtf8 . chunkToText) t)
+
+fromText :: T.Text -> Text
+fromText s | T.null s = mempty
+fromText s = go (Text (R.one (chunk s)))
+  where
+  go t | n > threshold  = go (take (n `div` 2) t) <> go (drop (n `div` 2) t)
+       | otherwise      = t
+       where n = size t
+
+pack :: String -> Text
+pack = fromText . T.pack
```

Comment:
Here as well, we have an opportunity to split the string into the proper size of chunks before `pack`ing it, IIRC splitting up text results in needing to copy the bytes of each split into completely new Text's; so splitting up the strings before packing them would hopefully avoid some of that overhead.

Not sure which functions we're most performance conscious about, so use your discretion on what's worth investigating 😄

Reply:
> IIRC splitting up text results in needing to copy the bytes of each split into completely new Text's; so splitting up the strings before packing them would hopefully avoid some of that overhead.

`Text` is represented with an underlying byte array and an offset and length into that array (hence why it exposes a [`copy`](https://hackage.haskell.org/package/text-1.2.4.1/docs/src/Data.Text.html#copy) function that copies just that slice to a fresh array). You can have multiple text values that share the same array, just with different offsets and lengths, and I would assume that like `Text.chunksOf` will not produce copies.

So doing the chunking on the string and then packing those would lose out on this sharing!

Reply:
Ah, gotcha! That makes sense.

So if I'm understanding that correctly, the sharing would likely improve performance on creation, since we only allocate a single array and then swap some offsets around later;

But now I'm curious about memory residency. If I `pack` a very long string into a `Text`, then I `take` a small portion of it and throw away the rest, I'm still paying the memory cost of the whole array.

Whereas if we chunk, and then pack, each chunk has its own array, so if we `take` a few chunks, we only pay the memory residency of the text we actually use.

It seems `Text.append` doesn't check if we're appending two contiguous texts from the same array, so if you `take` perform any operation which involves an append we'll lose the sharing anyhow, so maybe my point is moot here.

Thanks for the clarification, really appreciate it 🙌🏼

Reply:
Yes, sharing of the underlying array is sort of a double-edged sword if you just `take 1` and then throw out the rest. You could have various operations pre-emptively copy, but somehow that doesn't seem like the right call.

It's probably a good idea to expose a `Text.copy` builtin.